crates/ruff_server/src/lint.rs

```diff
@@ -159,51 +160,27 @@ pub(crate) fn check(
             .or_default();
     }
 
-    let lsp_diagnostics = diagnostics
-        .into_iter()
-        .zip(noqa_edits)
-        .map(|(diagnostic, noqa_edit)| {
-            to_lsp_diagnostic(
-                diagnostic,
-                noqa_edit,
-                &source_kind,
-                locator.to_index(),
-                encoding,
-            )
-        });
-
-    let unsupported_syntax_errors = if settings.linter.preview.is_enabled() {
-        parsed.unsupported_syntax_errors()
-    } else {
-        &[]
-    };
-
-    let lsp_diagnostics = lsp_diagnostics.chain(
-        show_syntax_errors
-            .then(|| {
-                parsed
-                    .errors()
-                    .iter()
-                    .map(|parse_error| {
-                        parse_error_to_lsp_diagnostic(
-                            parse_error,
-                            &source_kind,
-                            locator.to_index(),
-                            encoding,
-                        )
-                    })
-                    .chain(unsupported_syntax_errors.iter().map(|error| {
-                        unsupported_syntax_error_to_lsp_diagnostic(
-                            error,
-                            &source_kind,
-                            locator.to_index(),
-                            encoding,
-                        )
-                    }))
-            })
+    let lsp_diagnostics =
+        messages
             .into_iter()
-            .flatten(),
-    );
+            .zip(noqa_edits)
+            .filter_map(|(message, noqa_edit)| match message {
+                Message::Diagnostic(diagnostic_message) => Some(to_lsp_diagnostic(
+                    diagnostic_message,
+                    noqa_edit,
+                    &source_kind,
+                    locator.to_index(),
+                    encoding,
+                )),
+                Message::SyntaxError(syntax_error_message) => show_syntax_errors.then(|| {
+                    syntax_error_to_lsp_diagnostic(
+                        syntax_error_message,
+                        &source_kind,
+                        locator.to_index(),
+                        encoding,
+                    )
+                }),
```

Comment:
nit: I think now I'd prefer to expand the condition for `show_syntax_errors`:
```rs
if show_syntax_errors {
    Some(...)
} else {
    None
}
```
I mainly avoided that previously because it was part of a `chain` which I think is now removed.